lib/spack/spack/database.py

```diff
@@ -374,9 +374,29 @@ def _read_suppress_error():
             old_data = self._data
             try:
                 self._data = {}
+                processed_specs = set()
+                for key, entry in old_data.items():
+                    try:
+                        layout = spack.store.layout
+                        if entry.spec.external:
+                            layout = None
+                        kwargs = {
+                            'spec': entry.spec,
+                            'directory_layout': layout,
+                            'explicit': entry.explicit
+                        }
+                        self._add(**kwargs)
+                        processed_specs.add(entry.spec)
+                    except Exception as e:
+                        # Something went wrong, so the spec was not restored
+                        # from old data
+                        tty.debug(e.message)
+                        pass
```

Comment:
@alalazo: Ok last question on this PR: why does the new code above prefer data in the old DB to data from `spec.yaml` files?  I think the `spec.yaml` files should be considered authoritative.  Is this an artifact of the way this PR started -- with external stored only in the DB and not in the spec?  Could we just keep the old logic?

The logic above is "correct" but I think it does things the wrong way around for reindex.  In particular, reindex is supposed to restore a database from `spec.yaml` files, in case the DB is corrupted or otherwise inaccurate.  

The logic above reads the old DB and adds what it finds *in the old DB file* to the new DB.  It then subtracts the already-read specs from what is available in `spec.yaml` files (via `directory_layout`) and adds the remainder.  We do catch `Exception` above in case anything goes wrong while parsing a spec from the old DB (e.g., if it is corrupted) but this may not be the only thing wrong with the old DB.  The old DB may be corrupted in a valid, readable way, but this would prefer old DB data to data from `spec.yaml` files.  Also, if the hashes in the old DB differ somehow from what is on disk, the algorithm here will perpetuate the old, incorrect hashes, along with any read in from `spec.yaml` files on disk.

Consider also a use case where someone moves their Spack installation and runs `sed` on all the `spec.yaml` files.  I think they should be able to reindex after that and restore a working database, assuming they created valid `spec.yaml` files.  The above logic prevents that.

Basically I think this algorithm should be like it was before: it should loop *first* over `directory_layout.all_specs()` and only *supplement* what it finds in `spec.yaml` with things like `explicit` from the DB.

Finally, note that this doesn't prevent us from finding missing specs.  The `spec.yaml` files store info on the root and all dependencies, and the DB will deduce when deps are missing.